docs/ADR-X-cardano-api-exports-convention.md

```diff
@@ -0,0 +1,94 @@
+
+# Context
+
+TODO
+
+# Proposed Design
+
+## Top level exports
+
+Each top level export should export distinct symbols.
+This may depend on the case, for example: `Cardano.Api.Byron` symbols which are still in use in current eras, can be reexported through `Cardano.Api`.
+
+* `Cardano.Api` - everything domain related for the currently used and upcoming eras
+
+* `Cardano.Api.<F>`, where `<F>` represents a feature.
+  An aggregate module providing one of core `cardano-api` functionalities.
+  The name has to be a singular noun.
+  The purpose of this export is to provide users a more granular control of imports if they don't want to import a catch-all `Cardano.Api` module.
+  For example:
+
+    * `Cardano.Api.Address`
+    * `Cardano.Api.Block`
+    * `Cardano.Api.Certificate`
+    * `Cardano.Api.Eon`
+    * `Cardano.Api.Genesis`
+    * `Cardano.Api.Governance`
+    * `Cardano.Api.IPC`
+    * `Cardano.Api.Key`
+    * `Cardano.Api.LedgerState`
+    * `Cardano.Api.Query`
+    * `Cardano.Api.Serialisation`
+    * `Cardano.Api.Tx`
+
+  The following rules apply:
+
+  1. The modules should not be very granular, only related to a particular feature set.
+     For example
+
+     * `Cardano.Api.Governance.Actions.ProposalProcedure` should not be a top level export and should be exported through `Cardano.Api.Governance`.
+     * `Cardano.Api.LedgerEvents` should not be a top level export and should be exported through `Cardano.Api.LedgerState`.
+
+  1. The module should be reexported through `Cardano.Api`.
+    This makes using `cardano-api` easier, by not forcing users to decide what to import, and allowing them to just simply write `import Cardano.Api`.
+
+* `Cardano.Api.Byron` - everything for the obsoleted Byron era
+
+* `Cardano.Api.Compatible` - everything for the older eras.
+  The purpose of this module to provide a limited backwards compatibility for eras not in use on Cardano mainnet.
+  This is meant only for supporting of tests which hardfork through eras.
+
+  For both `Cardano.Api.Byron` and `Cardano.Api.Compatible` the following rules apply:
+
+  1. Any symbol from previous era, not in use in current era, should be exported through the respective module for older eras.
+  2. Any symbol from previous era and still in use in the current era, should be exported through `Cardano.Api` and additionally through `Cardano.Api.(Byron|Compatible)`.
+
+* `Cardano.Api.Experimental` - a transient module for implementing new not yet stabilised APIs.
+  Not meant to be used by `cardan-api` users directly.
+
+* `Cardano.Api.<U>` utility module, where `<U>` represents a utilities set.
+  For example `Cardano.Api.Pretty`, `Cardano.Api.Monad.Error`.
+
+  * Reexports from upstream libraries like consensus, network or ledger, also fall into this category e.g. `Cardano.Api.Ledger`.
+
+  * Modules which are exposing general purpose classes and functions should go also into a separate module, and they don't need to be reexported through `Cardano.Api`.
```

Comment:
So this is the problem with an "everything" module like `Cardano.Api`, it can't be everything. I'd like to demarcate this via the module name more clearly. What about something like:
```
Cardano.Api.Direct.UTxO   -- Direct import modules (not re-exported)
```
or overall something like:
```
Cardano.Api/
  Core/       -- re-exported through Cardano.Api
  Direct/     -- must be imported directly
  Internal/   -- implementation details
```

Reply:
I think we only need in this case an `Internal` module. Everything what is exposed only  `Cardano.Api` should go to `Cardano.Api.Internal`. For example, if our current module `Cardano.Api.Internal.Fees` is meant to be only exposed through `Cardano.Api`, it should be where it's now.

If a module is meant to be a top level export, not necessarily exposed through `Cardano.Api`, it should be a top level module. If our current `Cardano.Api.Internal.Fees` is meant to be a top level export only, it should be split into two:
* `Cardano.Api.Fees` - public API
* `Cardano.Api.Fees.Internal` - internal functions, not exposed

I've described it in "Deeper level exports" section.